django_walletpass/classviews.py

```diff
@@ -37,7 +38,12 @@ def list(self, request, device_library_id, pass_type_id):
             return Response({}, status=status.HTTP_400_BAD_REQUEST)
 
         if 'passesUpdatedSince' in request.GET:
-            passes = passes.filter(updated_at__gt=datetime.strptime(request.GET['passesUpdatedSince'], FORMAT))
+            try:
+                dt = datetime.strptime(request.GET['passesUpdatedSince'], FORMAT)
```

Comment:
What values of `request.GET['passesUpdatedSince']` will trigger this bug?

Reply:
26th of March, 2-3 am in the morning, when clocks switched from CET to CEST.

Reply:
Payload:

<img width="454" alt="Screenshot 2023-03-29 at 08 42 54" src="https://user-images.githubusercontent.com/2073086/228448188-c01c6381-e1c3-4e91-a7d3-2fec23776e90.png">

stack:
<img width="628" alt="Screenshot 2023-03-29 at 08 43 49" src="https://user-images.githubusercontent.com/2073086/228448287-fcc95877-a92c-40fd-8335-b7fa51104887.png">

I wonder if loading with a timezone would help, but not sure which one to apply...

Reply:
as far as I understand, this is a tag we are sending to apple earlier (https://messangi.com/documentation/doku.php?id=restapi:wallet ), hence if we migrate to UTC, the error shouldn't happen. what do you think @alexandernst ?

Reply:
ping @alexandernst

Reply:
@NachE can you help us? 🙏

Reply:
According to documentation:
* https://developer.apple.com/documentation/walletpasses/get_the_list_of_updatable_passes
* https://developer.apple.com/documentation/walletpasses/serialnumbers

The requested value of `passesUpdatedSince` is the value of previously response containing `lastUpdated` (https://github.com/Develatio/django-walletpass/blob/2.0/django_walletpass/classviews.py#LL45C31-L45C31) so if the date does not contemplates the timezone it is completely our fault.

We should use UTC dates always. But at this point there is already some passes out there with no-utc dates and we should support legacy dates requested in `passesUpdatedSince` even if we always use utc in `lastUpdated` from now.

Reply:
so what should be the next steps, to get this either rejected or merged @alexandernst @NachE ?

Reply:
* keep this `FORMAT` (https://github.com/Develatio/django-walletpass/pull/14/files#diff-360a103edfe15129e14f804814d65d056e45cb73209df4571b633c3ce03af4c1R17) as-it for legacy support
* add new UTC format
* test `.GET['passesUpdatedSince']` (https://github.com/Develatio/django-walletpass/pull/14/files#diff-360a103edfe15129e14f804814d65d056e45cb73209df4571b633c3ce03af4c1R42) format against legacy `FORMAT` and also against new UTC format ( try/except ?)
* if the format match against legacy FORMAT, convert to UTC-1day
* In the response (https://github.com/Develatio/django-walletpass/blob/2.0/django_walletpass/classviews.py#LL45C16-L45C16) always use new UTC format

Doing this, those clients who has stored the non-UTC format will receive the new UTC format and in the next request the new UTC format will be used.

Reply:
We should not force to set `USE_TZ`. In fact, this patch allows you to use _utc_ or _timezone_ interchangeably. See this POC:

```python
import pytz
from datetime import datetime
from django.utils.timezone import make_aware
from dateutil.parser import parse

tz = pytz.timezone('Europe/Copenhagen')
tz.localize(parse('2013-03-31T02:30:00'), is_dst=None)
```

```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/pytz/tzinfo.py", line 341, in localize
    raise NonExistentTimeError(dt)
pytz.exceptions.NonExistentTimeError: 2013-03-31 02:30:00
```

Even with this patch, the original error is still thrown if `tz` is not `UTC`. But now this patch catches the `NonExistentTimeError` and transforms the date, so the timezone is irrelevant.

Reply:
I agree with this patch. I leave a prudential window of time so that @alexandernst  can do a review before proceeding with the merge.